### PR TITLE
fix: improve error message from snapd pack

### LINKED ISSUE #4499: Failures when running `snap pack` don't get sent to the user.
### Bug Description

If `snap pack` fails, the error message (which is often helpful in determining why it failed) is hidden unless you enable debug logging or look in the detailed log file. Instead, the error message shows the specific command that was run in the instance (which isn't particularly useful to the user).

### To Reproduce

Run `snapcraft pack` with the provided `snapcraft.yaml`.

### Environment

Kubuntu 23.04

### snapcraft.yaml

```shell
name: my-snap-name # you probably want to 'snapcraft register <name>'
base: core22 # the base snap is the execution environment for this snap
version: '0.1' # just for humans, typically '1.2+git' or '1.3.2'
summary: Single-line elevator pitch for your amazing snap # 79 char long summary
description: |
  This is my-snap's description. You have a paragraph or two to tell the
  most important story about your snap. Keep it under 100 words though,
  we live in tweetspace and your description wants to look good in the snap
  store.

grade: devel # must be 'stable' to release into candidate/stable channels
confinement: devmode # use 'strict' once you have the right plugs and slots

apps:
  thing:
    command: this/file/does/not/exist

parts:
  my-part:
    # See 'snapcraft plugins'
    plugin: nil
```

### Relevant log output

```shell
lengau@ratel:~/tmp$ snapcraft pack
Launching instance...                                                                                                                      
Executed: pull my-part                                                                                                                     
Executed: build my-part                                                                                                                    
Executed: stage my-part                                                                                                                    
Executed: prime my-part                                                                                                                    
Executed parts lifecycle                                                                                                                   
Generated snap metadata                                                                                                                    
Command '['snap', 'pack', '--filename', 'my-snap-name_0.1_amd64.snap', '--compression', 'xz', PosixPath('/root/prime'), PosixPath('/root/project')]' returned non-zero exit status 1.                                                                                                  
Failed to execute pack in instance.                                                                                                        
Recommended resolution: Run the same command again with --debug to shell into the environment if you wish to introspect this failure.      
Full execution log: '/home/lengau/.local/state/snapcraft/log/snapcraft-20231215-111314.031750.log'
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sergiusens**
- [ ] Have you followed the [guidelines for contributing](https://github.com/canonical/snapcraft/blob/main/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----
Fixes #4499

### CODE REVIEW

File: requirements-devel.txt [hunk 1]
```diff
@@ -51,7 +51,7 @@ craft-archives==2.0.1
     # via
     #   craft-application
     #   snapcraft (setup.py)
-craft-cli==2.12.0
+craft-cli==2.13.0
```
Comment (sergiusens) [inline requirements-devel.txt:54]: release that shows details

File: tests/unit/test_errors.py [hunk 2]
```diff
@@ -40,3 +43,86 @@ def test_maintenance_base_fallback():
...
+
+
+def test_snap_pack_error():
+    error = errors.SnapPackError(
+        call_error=subprocess.CalledProcessError(
+            returncode=1,
+            cmd=["snap", "pack"],
+            stderr=textwrap.dedent(
+                """\
+                2024/12/20 11:25:37.751687 container.go:411: in snap "my-snap-name": path "this/file/does/not" does not exist
```
Comment (lengau) [inline tests/unit/test_errors.py:55]: It's so beautiful

File: tests/unit/test_errors.py [hunk 3]
```diff
@@ -40,3 +43,86 @@ def test_maintenance_base_fallback():
...
+                2024/12/20 11:25:37.751697 container.go:411: in snap "my-snap-name": path "this/file" does not exist
+                2024/12/20 11:25:37.751700 container.go:411: in snap "my-snap-name": path "this" does not exist
+                error: referenced command not found in snap
+                """
+            ),
+        )
+    )
+
+    assert str(error) == "Snapd failed to pack"
+    assert error.details == "referenced command not found in snap"
```
Comment (lengau) [inline tests/unit/test_errors.py:90]: It's beautiful!

**Review by lengau — APPROVED** [review]
I am so happy about this! Thank you!

File: snapcraft/errors.py [hunk 4]
```diff
@@ -180,3 +181,25 @@ class SnapcraftAssertionError(SnapcraftError):
 
     Not to be confused with Python's built-in AssertionError.
     """
+
+
+class SnapPackError(SnapcraftError):
+    """Snapd packing error."""
+
+    def _get_error_string_from_stderr(self, stderr: str) -> str | None:
```
Comment (bepri) [inline snapcraft/errors.py:189]:
```suggestion
    def _get_error_string_from_stderr(self, stderr: str | None) -> str | None:
```

My IDE reports the first `if` is unreachable without this

**Review by bepri — APPROVED** [review]
Nice! Just a note on typing.

**Review by mr-cal — APPROVED** [review]
Tested, looks great!

**Pull request merged**